src/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs

```diff
@@ -20,28 +20,38 @@ internal partial class DiagnosticAnalyzerService
     private partial class DiagnosticIncrementalAnalyzer
     {
         /// <summary>
-        /// Cached data from a real <see cref="Project"/> instance to the cached diagnostic data produced by
+        /// Cached data from a real <see cref="ProjectState"/> instance to the cached diagnostic data produced by
         /// <em>all</em> the analyzers for the project.  This data can then be used by <see
         /// cref="DiagnosticGetter.ProduceDiagnosticsAsync"/> to speed up subsequent calls through the normal <see
         /// cref="IDiagnosticAnalyzerService"/> entry points as long as the project hasn't changed at all.
         /// </summary>
-        private readonly ConditionalWeakTable<Project, StrongBox<(ImmutableArray<DiagnosticAnalyzer> analyzers, ImmutableDictionary<DiagnosticAnalyzer, DiagnosticAnalysisResult> diagnosticAnalysisResults)>> _projectToForceAnalysisData = new();
+        /// <remarks>
+        /// This table is keyed off of <see cref="ProjectState"/> but stores data from <see cref="SolutionState"/> on
+        /// it.  Specifically <see cref="SolutionState.Analyzers"/>.  Normally keying off a ProjectState would not be ok
+        /// as the ProjectState might stay the same while the SolutionState changed.  However, that can't happen as
+        /// SolutionState has the data for Analyzers computed prior to Projects being added, and then never changes.
+        /// Practically, solution analyzers are the core Roslyn analyzers themselves we distribute, or analyzers shipped
+        /// by vsix (not nuget).  These analyzers do not get loaded after changing *until* VS restarts.
+        /// </remarks>
+        private readonly ConditionalWeakTable<ProjectState, StrongBox<(ImmutableArray<DiagnosticAnalyzer> analyzers, ImmutableDictionary<DiagnosticAnalyzer, DiagnosticAnalysisResult> diagnosticAnalysisResults)>> _projectToForceAnalysisData = new();
 
         public async Task<ImmutableArray<DiagnosticData>> ForceAnalyzeProjectAsync(Project project, CancellationToken cancellationToken)
         {
+            var projectState = project.State;
+
             try
             {
-                if (!_projectToForceAnalysisData.TryGetValue(project, out var box))
+                if (!_projectToForceAnalysisData.TryGetValue(projectState, out var box))
                 {
                     box = new(await ComputeForceAnalyzeProjectAsync().ConfigureAwait(false));
 
                     // Try to add the new computed data to the CWT.  But use any existing value that another thread
                     // might have beaten us to storing in it.
 #if NET
-                    _projectToForceAnalysisData.TryAdd(project, box);
-                    Contract.ThrowIfFalse(_projectToForceAnalysisData.TryGetValue(project, out box));
+                    _projectToForceAnalysisData.TryAdd(projectState, box);
+                    Contract.ThrowIfFalse(_projectToForceAnalysisData.TryGetValue(projectState, out box));
```

Comment:
>Contract.ThrowIfFalse(_projectToForceAnalysisData.TryGetValue(projectState, out box)); [](http://example.com/codeflow?start=20&length=86)

nit: seems like this could limited to only be done if the TryAdd returned false.

Reply:
Yeah. I didn't care much :-)

Reply:
Changed to only do this if we can't get the value.